### DISCUSSION TITLE
Is there a way to prevent casting the extension context as any (e.g. `await (context as any).findFirst({ where })`)?

### DISCUSSION THREAD

**Discussion by alexwork1611** [prisma/prisma#25476] [Q&A] [answered]
### Question

I'd like to know if there's a way to properly type the context inside Prisma extensions. See example below (taken from https://www.prisma.io/docs/orm/prisma-client/client-extensions/model):

![image](https://github.com/user-attachments/assets/e0545ddf-76f8-44ba-a2e2-a950a05bbdbb)

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment, marked as answer]
Hi @alexwork1611 👋 

The challenge with typing the context inside Prisma extensions is a known limitation due to how TypeScript handles generics and the `this` keyword. When you apply a generic type `T` to `this`, TypeScript cannot automatically infer the type of `this`, making it difficult to type the `context` in a type-safe manner.

Unfortunately, this is a TypeScript limitation, and it's nearly impossible to make this type-safe when applying it to `$allModels`. In this case, the recommended approach is to use a less type-safe approach within the extension while maintaining a fully type-safe extension API. This means using `any` for the context and ensuring that the methods exposed by the extension are type-safe.

You can also look at this open [issue](https://github.com/prisma/prisma/issues/17991).

**Reply by alexwork1611** [reply]
Hello, Mr. Etim! That's very clear. Thank you so much!

**Reply by ludralph** [reply]
You're welcome 🙏

**Reply by alexwork1611** [reply]
So I came up with something, but I'm stuck:

```typescript
import { Prisma, PrismaClient } from '@prisma/client'

type PrismaClientModelKeys = Exclude<
  keyof PrismaClient extends infer TPrismaClientKey
    ? TPrismaClientKey extends `$${string}`
      ? never
      : TPrismaClientKey
    : never,
  symbol
>

type PrismaModel = PrismaClient[PrismaClientModelKeys]

export const UserModel = Prisma.defineExtension({
  model: {
    $allModels: {
      async exists<T extends PrismaModel>(
        this: T,
        where: NonNullable<Prisma.Args<T, 'findFirst'>['where']>,
      ) {
        const context = Prisma.getExtensionContext(this)

        if (context.findFirst) {
          const result = await context.findFirst({ where })

          return result !== null
        }
      },
    },
  },
})
```

I get an error at `findFirst`:

![image](https://github.com/user-attachments/assets/52c99f9a-5926-449d-b86d-1853ad4c6097)

Here's my Prisma schema:

```
// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = env("DATABASE_URL")
}

model User {
  email    String @id @unique
  password String
  Post     Post[]
}

model Post {
  id       String  @id @default(cuid())
  title    String
  content  String?
  authorId String
  author   User    @relation(fields: [authorId], references: [email])
}
```

Do you have any ideas, sir?
